### DISCUSSION TITLE
Many-Objective and Iteration Evaluation

### DISCUSSION THREAD

**Discussion by faranak1991** [anyoptimization/pymoo#561] [General]

**Comment by faranak1991** [comment]
Hello everyone,

I have been using the following commands to minimize the 5-objective problem using both NSGA3 (and NSGA2). I am tracking the best value of each function over iterations. However, I have noticed that the best value of objective does not seem to consider all previous generations when comparing. I expected the best value of each objective to show a downward trend. It shows that the best value after some generations appears to restart. Here's an example:

Generation 1: Best_F: [0.00023328   0.00086596    0.00054079     0.01242588     0.01710647]
Generation 2: Best_F: [0.00023328   0.00086596    0.00054079     0.0146338      0.00846329]
Generation 3: Best_F: [0.00020456   0.00086524    0.00192541      0.01354562    0.00846329]
Generation 4: Best_F: [0.00042178    0.00084673    0.00321917      0.0097553      0.00483079]
Generation 5: Best_F: [0.00042178.   0.00084673    0.00279974     0.0097553      0.00483079]
Generation 6: Best_F: [0.0002391      0.00084328    0.00075635    0.0097553      0.00810061]
Generation 7: Best_F: [0.00022657    0.00088079    0.00051862    0.0097553       0.01031505]
Generation 8: Best_F: [0.00022779    0.00081578     0.00037185    0.01208127     0.01104479]

Is this behavior normal? I would appreciate it if someone could help figure this out. Any assistance would be greatly appreciated.




a part of my code:
class myProblem(Problem):
    def __init__(self):
        super().__init__(n_var=22, n_obj=5, n_constr=0, xl=lower_bounds, xu=upper_bounds)
    def _evaluate(self, X, out, *args, **kwargs):
        n_samples = X.shape[0]         
        f1 = np.empty((n_samples, 1))
        f2 = np.empty((n_samples, 1))
        f3 = np.empty((n_samples, 1))
        f4 = np.empty((n_samples, 1))
        f5 = np.empty((n_samples, 1))
        for i in range(n_samples):
            x = X[i, :]  # Extract the i-th individual's parameters
            objectives = objective_function(x)  
            # Assign the computed objectives to fi
            f1[i, 0] = objectives[0]  #  objectives[0] is delta1
            f2[i, 0] = objectives[1]  #  objectives[1] is delta2
            f3[i, 0] = objectives[2]  #  objectives[2] is delta3
            f4[i, 0] = objectives[3]  #  objectives[3] is delta4
            f5[i, 0] = objectives[4]  #  objectives[4] is delta5
        out["F"] = np.column_stack((f1, f2, f3, f4, f5))

class MyOutput(Output):
    def __init__(self):
        super().__init__()
        self.F_mean = Column("F_mean", width=13)
        self.F_std = Column("F_std", width=13)
        self.best_F = Column("Best_F", width=13) 
        self.columns += [self.F_mean, self.F_std, self.best_F]  # Include the new column
    def update(self, algorithm):
        super().update(algorithm)  
        F_values = algorithm.pop.get("F") # containing the objective function values for each solution in the population
        self.F_mean.set(np.mean(F_values))
        self.F_std.set(np.std(F_values))
        # Calculate the best F value
        best_F_value = np.min(F_values, axis=0)
        self.best_F.set(best_F_value)
               
          # Write the data to the output file
        with open('output.txt', 'a') as output_file:
            #output_file.write(f"F_mean: {np.mean(F_values)}\n")
            output_file.write(f"Best_F: {best_F_value}\n")
            output_file.write("-" * 30 + '\n')

**Comment by blankjul** [comment]
your code is not easy to follow. but are you basically talking about the ideal point in your population?

**Reply by faranak1991** [reply]
Yes.
